api

```diff
@@ -1091,7 +1091,11 @@ refresh_all_pkgapp_status() { #for every package-app, if dpkg thinks it's instal
   #repeat for every package-type app
   local IFS=$'\n'
   for app in $(list_apps package) ;do
-    refresh_pkgapp_status "$app"
+    refresh_pkgapp_status "$app" &  declare PID_$(TMP=${app//-/_};echo ${TMP// /_})=$!
+  done
+  for app in $(list_apps package) ;do
+    varname=PID_$(TMP=${app//-/_};echo ${TMP// /_})
+    wait $(echo ${!varname})
```

Comment:
This seems overdesigned. On my system it took 4.2 seconds.
The single-threaded version took 2.6 seconds on my system.
I found a more efficient way to do multithreading:
```bash
refresh_all_pkgapp_status() { #for every package-app, if dpkg thinks it's installed, then mark it as installed.
  #repeat for every package-type app
  local IFS=$'\n'
  local PIDS
  for app in $(list_apps package) ;do
    refresh_pkgapp_status "$app" &
    PIDS+="$!"$'\n'
  done
  for PID in $PIDS ;do
    wait $PID
  done
}
```

Reply:
Further optimized:
```bash
refresh_all_pkgapp_status() { #for every package-app, if dpkg thinks it's installed, then mark it as installed.
  #repeat for every package-type app
  local IFS=$'\n'
  local PIDS=()
  for app in $(list_apps package) ;do
    refresh_pkgapp_status "$app" &
    PIDS+=("$!")
  done
  wait "${PIDS[@]}" #wait for all subprocesses to complete
}
```

Reply:
I've gone ahead and improved the branch with my changes.

Reply:
> This seems overdesigned. On my system it took 4.2 seconds. The single-threaded version took 2.6 seconds on my system. I found a more efficient way to do multithreading:

did you report these numbers backward? you said the old method (where each refresh had to wait on another) took 2.6 seconds, and my new method took 4.2 seconds

Reply:
regardless, you method seems to work fine and is definitely a cleaner way of writing it

Reply:
> > This seems overdesigned. On my system it took 4.2 seconds. The single-threaded version took 2.6 seconds on my system. I found a more efficient way to do multithreading:
> 
> did you report these numbers backward? you said the old method (where each refresh had to wait on another) took 2.6 seconds, and my new method took 4.2 seconds

Yes, your multithreaded method actually took 4.2 seconds. This could be attributed to me testing your multithreaded version first, then the original afterwards. The first time could have been a "cold-start" and the dpkg file first needed to be read into cache or something.
Subsequent tests of your method are 2.2 seconds on average, where my multithreaded method is 1.5 on average, on my system.